internal/bgp/frr/config.go

```diff
@@ -158,26 +162,55 @@ var reloadConfig = func() error {
 // generateAndReloadConfigFile takes a 'struct frrConfig' and, using a template,
 // generates and writes a valid FRR configuration file. If this completes
 // successfully it will also force FRR to reload that configuration file.
-func generateAndReloadConfigFile(config *frrConfig) error {
+func generateAndReloadConfigFile(config *frrConfig, l log.Logger) {
 	filename, found := os.LookupEnv("FRR_CONFIG_FILE")
 	if found {
 		configFileName = filename
 	}
 
 	configString, err := templateConfig(config)
 	if err != nil {
-		return err
+		level.Error(l).Log("op", "reload", "error", err, "cause", "template", "msg", "failed to reload config", "config", config)
+		return
 	}
 
 	err = writeConfig(configString, configFileName)
 	if err != nil {
-		return err
+		level.Error(l).Log("op", "reload", "error", err, "cause", "writeConfig", "msg", "failed to reload config", "config", config)
+		return
 	}
 
 	err = reloadConfig()
 	if err != nil {
-		return err
+		level.Error(l).Log("op", "reload", "error", err, "cause", "reload", "msg", "failed to reload config", "config", config)
+		return
 	}
 
-	return nil
+	level.Info(l).Log("op", "reload", "success", "reloaded config")
+}
+
+func debouncer(body func(config *frrConfig),
+	reload <-chan *frrConfig,
+	reloadInterval time.Duration) {
+	go func() {
+		var config *frrConfig
+		var timeOut <-chan time.Time
+		timerSet := false
+		for {
+			select {
+			case newCfg, ok := <-reload:
+				if !ok { // the channel was closed
+					return
+				}
+				config = newCfg
+				if !timerSet {
+					timeOut = time.After(reloadInterval)
+					timerSet = true
+				}
+			case <-timeOut:
+				body(config)
+				timerSet = false
```

Comment:
do u need to check config is failed any chance ever to get time out and somehow we dropped reload signal ?

Reply:
Since we decouple the calling site from the config execution, we can't really return any error.
This is not making things worse, because we already have the real reload decoupled by the go code (what we do is just call SIGUP). 
The only feedback we get is about what happens to the signal (which is unlikely to go wrong). Given the asyncrhonousness of the call, I preferred embedding the logger in the function and logging from there. That's why `body` does not return any error.

Reply:
What should the user do in the(unlikely) event where something did go wrong along the way?

Reply:
I guess it depends on the nature of the error, and not dependant on this PR.
We already detach the request for a reload to the real reload, so there's already no way to understand why it failed.

Reply:
Filed https://github.com/metallb/metallb/issues/1046 to keep track of this.

Reply:
I was thinking about if something wrong happens before we reach the SIGHUP part (writeConfig fails for some reason/etc..)

Reply:
Not sure I follow. We log for that in `generateAndReloadConfigFile`.
One other thing we could _ideally_ do is to retrigger the timeout in case of failure, not sure it's worth the effort. I suspect if something fails in that context, will continue to fail.

Reply:
Thanks for explaining offline, I thought that in the current implementation we retry in case of an error but turns out we only log:
https://github.com/metallb/metallb/blob/f8e56adaf2d8a84f3d32e6a869fd8e31c5d06c3c/speaker/main.go#L367